runtime/src/accounts.rs

```diff
@@ -320,34 +329,56 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, (account, None));
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &mut HashMap<Pubkey, Arc<AtomicU64>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
-    ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+    ) -> Result<CreditOnlyLocks> {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k)
+                || (credit_only_locks.contains_key(k)
+                    && Arc::strong_count(credit_only_locks.get(&k).unwrap()) > 1)
```

Comment:
strong_count() is designed for a very specific use case, and is prone to a host of correctness issues otherwise @sagar-solana. I would prefer if we used something like a Mutex<usize> to track the number of credit-only-threads for a particular account. And then:

For credit-debit transactions:
1) Grab the credit-only Mutex, check that it's not zero
2) Add the credit-debit to the `locks` list
3) Release the credit-only Mutex

For credit-only accounts
1) Grab the credit-only Mutex<usize>
2) Check the `locks` queue doesn't contain this account
3) Increment the credit-only Mutex<usize>, release the lock

Reply:
To be clear, you are talking about a Mutex wrapping a counter integer, stored in the Accounts > credit_only_account_locks HashMap?

Reply:
yes i think that's the `credit_only_locks` HashMap above which I think should be a Arc<RwLock<HashMap<Pubkey, Arc<Mutex<usize>>>>. Then instead of passing the `credit_only_locks` as a `mut&` (because that means we have to lock the entire hashmap), we can just pass in a reference to the HashMap. In this `lock_account` function we  most of the time we only need  read lock to grab a read a copy of the Mutex ref-counter for a credit-only account,and then release the lock.  When we need to make a new entry in the HashMap, we grab a write lock, insert, then drop the lock.

Reply:
We still need the AtomicU64 to collect the actual credits...
So: `Arc<RwLock<HashMap<Pubkey, (Arc<AtomicU64>, Arc<Mutex<u64>>)>>`? That's a mouthful

Reply:
Hehe yeah we can make it a struct haha